streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/InMemoryTimeOrderedKeyValueBuffer.java

```diff
@@ -0,0 +1,130 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.kstream.internals.suppress;
+
+import org.apache.kafka.streams.KeyValue;
+
+import java.util.ConcurrentModificationException;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.TreeMap;
+
+class InMemoryTimeOrderedKeyValueBuffer<K, V> implements TimeOrderedKeyValueBuffer<K, V> {
+    private long memBufferSize = 0;
+    private final Map<K, TimeKey<K>> index = new HashMap<>();
+    private final TreeMap<TimeKey<K>, SizedValue<V>> sortedMap = new TreeMap<>();
+
+    private static class SizedValue<V> {
+        private final V v;
+        private final long size;
+
+        private SizedValue(final V v, final long size) {
+            this.v = v;
+            this.size = size;
+        }
+
+        public V value() {
+            return v;
+        }
+
+        public long size() {
+            return size;
+        }
+
+        @Override
+        public String toString() {
+            return "SizedValue{v=" + v + ", size=" + size + '}';
+        }
+    }
+
+    private class InnerIterator implements Iterator<KeyValue<TimeKey<K>, V>> {
+        private final Iterator<Map.Entry<TimeKey<K>, SizedValue<V>>> delegate = sortedMap.entrySet().iterator();
+        private Map.Entry<TimeKey<K>, SizedValue<V>> last;
+
+        @Override
+        public boolean hasNext() {
+            return delegate.hasNext();
+        }
+
+        @Override
+        public KeyValue<TimeKey<K>, V> next() {
+            final Map.Entry<TimeKey<K>, SizedValue<V>> next = delegate.next();
+            this.last = next;
+            return new KeyValue<>(next.getKey(), next.getValue().value());
+        }
+
+        @Override
+        public void remove() {
+            delegate.remove();
+            index.remove(last.getKey().key());
+            memBufferSize = memBufferSize - last.getValue().size();
+        }
+    }
+
+    @Override
+    public Iterator<KeyValue<TimeKey<K>, V>> iterator() {
+        return new InnerIterator();
+    }
+
+    @Override
+    public void put(final long time,
+                    final K key,
+                    final V value,
+                    final long recordSize) {
+        final TimeKey<K> previousKey = index.get(key);
+        // non-resetting semantics:
```

Comment:
This could be configurable in the future, but for now, we enforce the time limit in the following fashion:
* start a timer when a key first enters the buffer
* that key and its most recent value will be emitted when the time limit expires, regardless of how recently it has been updated

The primary advantage of this implementation is that we guarantee that if you set a 5-minute limit, we delay emitting the key for no more than five minutes. If we instead re-set the timer on each update, you might never see a record that gets consistently updated more frequently than the time limit.

My judgement was that this is the most intuitive default and starting point for the feature. If people want to configure it, we can easily add that option later.

Reply:
I agree with the described semantics.